comb_spec_searcher/limited_strategy_db.py

```diff
@@ -0,0 +1,149 @@
+"""
+    An alternative ruledb that permits restricting the number of applications of a given
+    set of strategies to a given limit.
+"""
+from copy import deepcopy
+from itertools import combinations
+from time import time
+from typing import Dict, Set, Tuple, Type
+
+from logzero import logger
+
+from comb_spec_searcher.rule_db import RuleDB
+from comb_spec_searcher.strategies import AbstractStrategy
+
+from .exception import SpecificationNotFound
+from .tree_searcher import (
+    Node,
+    iterative_proof_tree_finder,
+    iterative_prune,
+    prune,
+    random_proof_tree,
+)
+
+LabelRule = Tuple[int, Tuple[int, ...]]
+RulesDict = Dict[int, Set[Tuple[int, ...]]]
+
+
+class LimitedStrategyRuleDB(RuleDB):
+    def __init__(
+        self,
+        strategies_to_limit: Set[Type[AbstractStrategy]],
+        limit: int,
+        mark_verified: bool,
+        minimization_time_limit: int,
+    ) -> None:
+        self.strategies_to_limit = strategies_to_limit
+        self.limit = limit
+        self.mark_verified = mark_verified
+        self.minimization_time_limit = minimization_time_limit
+        super().__init__()
+
+    def get_eqv_rules_using_strategies(self) -> Set[LabelRule]:
+        """
+        returns a set of rules, up to equivalence, that come from
+        <self.strategies_to_limit>
+        """
+        eqv_rules_using_strategies: Set[LabelRule] = set()
+        for label_rule in self:
+            rule_strat = self.rule_to_strategy[label_rule]
+            start, ends = label_rule
+            if any(isinstance(rule_strat, strat) for strat in self.strategies_to_limit):
+                temp_start = self.equivdb[start]
+                temp_ends = tuple(sorted(map(self.equivdb.__getitem__, ends)))
+                eqv_rules_using_strategies.add((temp_start, temp_ends))
+        return eqv_rules_using_strategies
+
+    def find_specification(self, label: int, iterative: bool = False) -> Node:
+        """Search for a specification based on current data found."""
+        rules_dict = self.rules_up_to_equivalence()
+        # Prune all unverified labels (recursively)
+        # Afterward, only verified labels remain in rules_dict. In particular, there
+        # is a specification if <label> is in the rules_dict
+        if iterative:
+            rules_dict = iterative_prune(rules_dict, root=label)
+        else:
+            prune(rules_dict)  # this function removes rules not in a specification.
+
+        if self.mark_verified:
+            for ver_label in rules_dict.keys():
+                self.set_verified(ver_label)
```

Comment:
This should probably happen on line 103 (after `prune(temp_rules_dict)`) as otherwise you will verify things in a specification using more than the limit number of strategies.

Reply:
Agree. Fixed.